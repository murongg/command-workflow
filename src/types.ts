export interface StepTags {
  [x: string]: (() => string) | string
}
export interface Step {
  command: string
  error?: (error: Error) => void
  tags?: StepTags
  before?: (command: string, tags: Record<string, any>) => string | undefined
  after?: (command: string, buffer: Buffer) => void
}

