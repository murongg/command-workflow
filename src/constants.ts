export const DEFAULT_CONFIG_SUFFIXES = [
  '.config.js',
  '.config.mjs',
  '.config.ts',
  '.config.cjs',
  '.config.mts',
  '.config.cts',
]

export const DEFAULT_CONFIG_PREFIXES = [
  'cwf',
  'command-workflow',
  'command',
  'workflow',
]

export const getDefaultConfigPrefixes = () => {
  const files = []
  for (const prefix of DEFAULT_CONFIG_PREFIXES) {
    for (const suffix of DEFAULT_CONFIG_SUFFIXES)
      files.push(`${prefix}${suffix}`)
  }
  return files
}
