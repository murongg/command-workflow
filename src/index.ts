export { defineConfig } from './config'
export { start } from './run'
export type { StepTags, Step } from './types'
export type { UserConfig, UserConfigFn, UserConfigMap, UserConfigExport } from './config'
