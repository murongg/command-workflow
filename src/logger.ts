/* eslint no-console: 0 */
// fork vite logger.ts
import colors from 'picocolors'

export type LogType = 'error' | 'warn' | 'info'
export type LogLevel = LogType | 'silent'
export interface Logger {
  info(msg: string, options?: LogOptions): void
  warn(msg: string, options?: LogOptions): void
  warnOnce(msg: string, options?: LogOptions): void
  error(msg: string, options?: LogErrorOptions): void
  hasWarned: boolean
}

export interface LogOptions {
  clear?: boolean
  timestamp?: boolean
}

export interface LogErrorOptions extends LogOptions {
  error?: Error | null
}

export const LogLevels: Record<LogLevel, number> = {
  silent: 0,
  error: 1,
  warn: 2,
  info: 3,
}

export interface LoggerOptions {
  prefix?: string
}

export function createLogger(
  level: LogLevel = 'info',
  options: LoggerOptions = {},
): Logger {
  const { prefix = '[CWF]' } = options
  const thresh = LogLevels[level]

  function output(type: LogType, msg: string, options: LogErrorOptions = {}) {
    if (thresh >= LogLevels[type]) {
      const method = type === 'info' ? 'log' : type
      const format = () => {
        if (options.timestamp) {
          const tag
            = type === 'info'
              ? colors.cyan(colors.bold(prefix))
              : type === 'warn'
                ? colors.yellow(colors.bold(prefix))
                : colors.red(colors.bold(prefix))
          return `${colors.dim(new Date().toLocaleTimeString())} ${tag} ${msg}`
        }
        else {
          return msg
        }
      }
      if (options.error)
        msg += `\n${options.error.stack}`

      console[method](format())
    }
  }

  const warnedMessages = new Set<string>()

  const logger: Logger = {
    hasWarned: false,
    info(msg, opts) {
      output('info', msg, opts)
    },
    warn(msg, opts) {
      logger.hasWarned = true
      output('warn', msg, opts)
    },
    warnOnce(msg, opts) {
      if (warnedMessages.has(msg))
        return
      logger.hasWarned = true
      output('warn', msg, opts)
      warnedMessages.add(msg)
    },
    error(msg, opts) {
      logger.hasWarned = true
      output('error', msg, opts)
    },
  }

  return logger
}
