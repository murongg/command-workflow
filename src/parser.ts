import type { StepTags } from './types'

/**
 * e.g. git tag v#{tag} -m v#{message}
 * parse to git tag v1.0.0 -m v1.0.0
 * @param template
 * @param tags
 * @returns
 */
export function parserTemplateTag(template: string, tags: StepTags): string {
  return template.replace(/#\{(\w+)\}/g, (_, key) => {
    const value = tags[key]
    if (typeof value === 'function')
      return value()

    return value
  })
}
