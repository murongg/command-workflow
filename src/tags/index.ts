import { getTimestamp } from './date'
import { getCurrentGitBranch, getCurrentGitCommit, getCurrentGitCommitMessage, getCurrentGitRepo, getCurrentGitRepoName, getCurrentGitRepoOwner, getCurrentGitRepoUrl, getCurrentGitTag, getGitUserEmail, getGitUserName } from './git'

export const TAGS_MAP = {
  timestamp: getTimestamp,
  current_git_branch: getCurrentGitBranch,
  current_git_commit: getCurrentGitCommit,
  current_git_commit_message: getCurrentGitCommitMessage,
  current_git_tag: getCurrentGitTag,
  current_git_repo: getCurrentGitRepo,
  current_git_repo_url: getCurrentGitRepoUrl,
  current_git_repo_name: getCurrentGitRepoName,
  current_git_repo_owner: getCurrentGitRepoOwner,
  git_user_name: getGitUserName,
  git_user_email: getGitUserEmail,
}
