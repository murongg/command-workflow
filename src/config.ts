import { resolve } from 'path'
import { existsSync } from 'fs'
import { getDefaultConfigPrefixes } from './constants'

export interface UserConfig {
  /**
   * Is output log.
   * @default true
   */
  log?: boolean
  /**
   * Log level.
   * @default 'info'
   */
  logLevel?: 'debug' | 'info' | 'warn' | 'error'

}

export function defineConfig(config: UserConfig): UserConfig
export function defineConfig(config: Promise<UserConfig>): Promise<UserConfig>
export function defineConfig(config: UserConfig | Promise<UserConfig>): Promise<UserConfig> | UserConfig {
  return config
}

export async function loadConfigFromFile(configFile?: string, configRoot: string = process.cwd()) {
  let resolvedPath = ''
  if (configFile)
    resolvedPath = resolve(configFile)
  const configFiles = getDefaultConfigPrefixes()

  for (const filename of configFiles) {
    const filePath = resolve(configRoot, filename)
    if (!existsSync(filePath))
      continue

    resolvedPath = filePath
    break
  }

  try {
    const config = await import(resolvedPath)
    return config.default
  }
  catch (error) {
    console.error(error)
    return null
  }
}
