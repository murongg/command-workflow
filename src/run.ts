import { execSync } from 'child_process'
import { loadConfigFromFile } from './config'
import { parserTemplateTag } from './parser'

export async function run() {
  const config = await loadConfigFromFile()
  if (!config?.steps)
    throw new Error('No steps found in config file')
  if (config && config.steps) {
    const commands: string[] = []
    for (const step of config.steps) {
      const cmd = parserTemplateTag(step.command, step.tags || {})
      commands.push(cmd)
      try {
        execSync(cmd, { stdio: 'inherit' })
      }
      catch (error) {
        console.error(error)
      }
    }
  }
}
