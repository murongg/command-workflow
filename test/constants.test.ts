import { describe, expect, it } from 'vitest'
import { getDefaultConfigPrefixes } from '../src/constants'

describe('constants', () => {
  it('test getDefaultConfigPrefixes', () => {
    const files = getDefaultConfigPrefixes()
    expect(files).toMatchSnapshot()
  })
})
