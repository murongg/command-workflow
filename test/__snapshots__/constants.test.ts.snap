// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`constants > test getDefaultConfigPrefixes 1`] = `
[
  "cwf.config.js",
  "cwf.config.mjs",
  "cwf.config.ts",
  "cwf.config.cjs",
  "cwf.config.mts",
  "cwf.config.cts",
  "command-workflow.config.js",
  "command-workflow.config.mjs",
  "command-workflow.config.ts",
  "command-workflow.config.cjs",
  "command-workflow.config.mts",
  "command-workflow.config.cts",
  "command.config.js",
  "command.config.mjs",
  "command.config.ts",
  "command.config.cjs",
  "command.config.mts",
  "command.config.cts",
  "wordflow.config.js",
  "wordflow.config.mjs",
  "wordflow.config.ts",
  "wordflow.config.cjs",
  "wordflow.config.mts",
  "wordflow.config.cts",
]
`;
